### PR TITLE
[experimental] Add full_outer_join_mt experimental function.

### PR DESCRIPTION
**Pull request by tpoterba**
This was remarkably easy.

### CODE REVIEW

File: hail/python/hail/expr/functions.py [hunk 1]
```diff
@@ -2933,16 +2934,24 @@ def zip_with_index(a):
...
+    >>> hl.eval(hl.zip_with_index(['A', 'B', 'C'], first=False))
+    [('A', 0), ('B', 1), ('C', 2)]
+
+
     Parameters
     ----------
     a : :class:`.ArrayExpression`
+    index_first: :obj:`bool`
+        If ``True``, the index is the first value of the element tuples. If
+        ``False``, the index is tthe second value.
```
Comment (danking) [inline hail/python/hail/expr/functions.py]: tthe -> the

Comment (tpoterba) [reply hail/python/hail/expr/functions.py]: I hate the current generation of macbook keyboards.

File: hail/python/hail/experimental/full_outer_join_mt.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
+import hail as hl
+
+
+def full_outer_join_mt(left: hl.MatrixTable, right: hl.MatrixTable) -> hl.MatrixTable:
```
Comment (danking) [inline hail/python/hail/experimental/full_outer_join_mt.py]: tests!

File: hail/python/hail/expr/functions.py (hunk 1 above)
Comment (tpoterba) [reply hail/python/hail/expr/functions.py]: done

File: hail/python/hail/experimental/full_outer_join_mt.py [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+                               lambda i: hl.range(0, s.right_indices.length()).map(
+                                   lambda j: hl.struct(k=s.k, left_index=s.left_indices[i],
+                                                       right_index=s.right_indices[j]))))
+                     .when(hl.is_defined(s.left_indices),
+                           s.left_indices.map(
+                               lambda elt: hl.struct(k=s.k, left_index=elt, right_index=hl.null('int32'))))
+                     .when(hl.is_defined(s.right_indices),
+                           s.right_indices.map(
+                               lambda elt: hl.struct(k=s.k, left_index=hl.null('int32'), right_index=elt)))
+                     .or_error('assertion error')))
```
Comment (danking) [inline hail/python/hail/experimental/full_outer_join_mt.py]: this seems a bit odd to me. if you have

```
   a  a
1 10 11
```

```
   a  a
1 20 21
```
What does `key_indices` contain? It looks to me like you'd have
```
[ {k=a, left_index=0, right_index=0}
, {k=a, left_index=0, right_index=1}
, {k=a, left_index=1, right_index=0}
, {k=a, left_index=1, right_index=1}
]
```

This seems more like cross product than join? I'd have expected you end up with as many columns as you had distinct keys in the input and `left_entries` is an array of the entries with that column key.

Comment (tpoterba) [reply hail/python/hail/experimental/full_outer_join_mt.py]: This does do a cross product (isn't that what our Table joins do?)

File: hail/python/hail/experimental/full_outer_join_mt.py (hunk 2 above)
Comment (tpoterba) [reply hail/python/hail/experimental/full_outer_join_mt.py]: `.________.`

File: hail/python/hail/experimental/full_outer_join_mt.py (hunk 3 above)
Comment (catoverdrive) [reply hail/python/hail/experimental/full_outer_join_mt.py]: yep! All of our (non-distinct) table joins do an outer product over the set of rows corresponding to a given key.

File: hail/python/hail/experimental/full_outer_join_mt.py (hunk 2 above)
Comment (catoverdrive) [reply hail/python/hail/experimental/full_outer_join_mt.py]: tests would be great. there's a lot of logic in this function. I think I'd be happy with a single test that had out-of-order col keys and some duplicate key values.

File: hail/python/hail/expr/functions.py [hunk 4]
```diff
@@ -2933,16 +2934,24 @@ def zip_with_index(a):
     >>> hl.eval(hl.zip_with_index(['A', 'B', 'C']))
     [(0, 'A'), (1, 'B'), (2, 'C')]
 
+    >>> hl.eval(hl.zip_with_index(['A', 'B', 'C'], first=False))
```
Comment (catoverdrive) [inline hail/python/hail/expr/functions.py]: `index_first`

(doctests are failing)

**Pull request merged**